README.md
[![Kwalitee status](https://cpants.cpanauthors.org/dist/CPANfile-Parse-PPI.png)](https://cpants.cpanauthors.org/dist/CPANfile-Parse-PPI)
[![GitHub issues](https://img.shields.io/github/issues/perlservices/CPANfile-Parse-PPI.svg)](https://github.com/perlservices/CPANfile-Parse-PPI/issues)
[![CPAN Cover Status](https://cpancoverbadge.perl-services.de/CPANfile-Parse-PPI-0.06)](https://cpancoverbadge.perl-services.de/CPANfile-Parse-PPI-0.06)
[![Cpan license](https://img.shields.io/cpan/l/CPANfile-Parse-PPI.svg)](https://metacpan.org/release/CPANfile-Parse-PPI)

# NAME

CPANfile::Parse::PPI - Parse _cpanfile_s with PPI

# VERSION

version 0.06

# SYNOPSIS

```perl
use v5.24;
use CPANfile::Parse::PPI;

my $path     = '/path/to/cpanfile';
my $cpanfile = CPANfile::Parse::PPI->new( $path );

# or
# my $cpanfile = CPANfile::Parse::PPI->new( \$content );

for my $module ( $cpanfile->modules->@* ) {
    my $stage   = $module->{stage}   ? " on $module->{stage}"            : '';
    my $feature = $module->{feature} ? " for feature $module->{feature}" : '';

    say sprintf "%s is %s",
         $module->{name}, $module->{type}, $stage, $feature;
}
```

# METHODS

## new

```perl
my $path     = '/path/to/cpanfile';
my $cpanfile = CPANfile::Parse::PPI->new( $path );

# or
my $content  = <<'CPANFILE';
requires "CPANfile::Parse::PPI" => 3.6;';
on build => sub {
    recommends "Dist::Zilla" => 4.0;
    requires "Test2" => 2.311;
}
feature 'sqlite', "SQLite Support" => sub {
    requires DBD::SQLite
}
CPANFILE

my $cpanfile = CPANfile::Parse::PPI->new( \$content );
```

# ATTRIBUTES

## meta

Returns information about mirrors and OS name - if given in the cpanfile

## modules

Returns a list of modules mentioned in the cpanfile ("perl" is skipped).
Each element is a hashref with these keys:

- name
- version
- type
- stage
- feature

```perl
use CPANfile::Parse::PPI;
use Data::Printer;

my $required = 'requires "CPANfile::Parse::PPI" => 3.6;';
my $cpanfile = CPANfile::Parse::PPI->new( \$required );

my $modules = $cpanfile->modules;
p $modules;

__DATA__
[
    [0] {
        name      "CPANfile::Parse::PPI",
        stage     "",
        type      "requires",
        version   3.6
    }
]
```

# LIMITATIONS

As this is a static parser, this module cannot handle dynamic
code like

```perl
for my $module (qw/
    IO::All
    Zydeco::Lite::App
/) {
    requires $module, '0';
}
```

This module warns when the required "module" doesn't look like
a package name.

You can make it die when you pass `-strict` to the module
when you load it:

```perl
use CPANfile::Parse::PPI -strict;
use Data::Printer;

my $required = do { local $/; <DATA> };
my $cpanfile = CPANfile::Parse::PPI->new( \$required );

my $modules = $cpanfile->modules;

__DATA__
for my $module (qw/
    IO::All
    Zydeco::Lite::App
/) {
    requires $module, '0';
}
```



# Development

The distribution is contained in a Git repository, so simply clone the
repository

```
$ git clone git://github.com/perlservices/CPANfile-Parse-PPI.git
```

and change into the newly-created directory.

```
$ cd CPANfile-Parse-PPI
```

The project uses [`Dist::Zilla`](https://metacpan.org/pod/Dist::Zilla) to
build the distribution, hence this will need to be installed before
continuing:

```
$ cpanm Dist::Zilla
```

To install the required prequisite packages, run the following set of
commands:

```
$ dzil authordeps --missing | cpanm
$ dzil listdeps --author --missing | cpanm
```

The distribution can be tested like so:

```
$ dzil test
```

To run the full set of tests (including author and release-process tests),
add the `--author` and `--release` options:

```
$ dzil test --author --release
```

# AUTHOR

Renee Baecker <reneeb@cpan.org>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2020 by Renee Baecker.

This is free software, licensed under:

```
The Artistic License 2.0 (GPL Compatible)
```


Changes
Revision history for CPANfile::Parse::PPI
=========================================

0.03  2021-01-03  06:20:00

  * Add  a 'strict' mode (see https://github.com/PerlServices/CPANfile-Parse-PPI/commit/58f049)

0.02  2020-12-05  07:15:00

  * removed Data::Printer from test script (thanks to CPANtesters)

0.01  2020-12-04  14:32:00

  * initial release


lib/CPANfile/Parse/PPI.pm
package CPANfile::Parse::PPI;

# ABSTRACT: Parse I<cpanfile>s with PPI

use strict;
use warnings;

use PPI;
use Moo;
use Carp qw(carp croak);

our $VERSION = '0.04';

my $strict;

has modules => (
    is  => 'ro',
    isa => sub {
        die if 'ARRAY' ne ref $_[0];
    }
);

=begin pod_coverage

=head2 BUILDARGS

=head2 import

=end pod_coverage

=cut

sub BUILDARGS {
    my ($class, $file_or_code) = @_;

    my @modules = _parse( $file_or_code );

    return {
        modules => \@modules,
    };
}

sub import {
    $strict = 1 if grep{ $_ eq '-strict' }@_;
}

sub _parse {
    my ($file_or_code) = @_;

    my $doc = PPI::Document->new( $file_or_code );
    my $requires = $doc->find(
        sub { 
            $_[1]->isa('PPI::Token::Word') and (
                $_[1]->content eq 'requires' ||
                $_[1]->content eq 'recommends'
             )
        }
    );

    return if !$requires;

    my @modules;

    REQUIRED:
    for my $required ( @{ $requires || [] } ) {
        my $value = $required->snext_sibling;

        my $type = $required->content;

        my $can_string = $value->can('string') ? 1 : 0;
        my $prereq     = $can_string ?
            $value->string :
            $value->content;

        next REQUIRED if $prereq eq 'perl';

        if (
            $value->isa('PPI::Token::Symbol') ||
            $prereq =~ m{\A[^A-Za-z]}
        ) {
            carp  'Cannot handle dynamic code' if !$strict;
            croak 'Cannot handle dynamic code' if $strict;

            next REQUIRED;
        }

        my $stage = '';

        my $parent_node = $value;

        PARENT:
        while ( 1 ) {
            $parent_node = $parent_node->parent;
            last PARENT if !$parent_node;
            last PARENT if $parent_node->isa('PPI::Document');

            if ( $parent_node->isa('PPI::Structure::Block') ) {
                $parent_node = $parent_node->parent;

                my ($on) = $parent_node->find_first( sub { $_[1]->isa('PPI::Token::Word') && $_[1]->content eq 'on' } );

                next PARENT if !$on;

                my $word = $on->snext_sibling; 
                $stage = $word->can('string') ? $word->string : $word->content;
                
                last PARENT;
            }
        }

        my $version = '';
        my $sibling = $value->snext_sibling;
        SIBLING:
        while ( 1 ) {
            last SIBLING if !$sibling;

            do { $sibling = $sibling->snext_sibling; next SIBLING } if !$sibling->isa('PPI::Token::Operator');

            my $value = $sibling->snext_sibling;
            $version = $value->can('string') ? $value->string : $value->content;

            last SIBLING;
        }

        push @modules, {
            name    => $prereq,
            version => $version,
            type    => $type,
            stage   => $stage,
        };
    }

    return @modules;
}

1;

=head1 SYNOPSIS

    use v5.24;
    use CPANfile::Parse::PPI;
    
    my $path     = '/path/to/cpanfile';
    my $cpanfile = CPANfile::Parse::PPI->new( $path );
    
    # or
    # my $cpanfile = CPANfile::Parse::PPI->new( \$content );
    
    for my $module ( $cpanfile->modules->@* ) {
        my $stage = "";
        $stage    = "on $module->{stage}" if $module->{stage};
        say sprintf "%s is %s", $module->{name}, $module->{type};
    }

=head1 METHODS

=head2 new

    my $path     = '/path/to/cpanfile';
    my $cpanfile = CPANfile::Parse::PPI->new( $path );
    
    # or
    my $content  = <<'CPANFILE';
    requires "CPANfile::Parse::PPI" => 3.6;';
    on build => sub {
        recommends "Dist::Zilla" => 4.0;
        requires "Test2" => 2.311;
    }
    CPANFILE

    my $cpanfile = CPANfile::Parse::PPI->new( \$content );

=head1 ATTRIBUTES

=head2 modules

Returns a list of modules mentioned in the cpanfile ("perl" is skipped).
Each element is a hashref with these keys:

=over 4

=item * name

=item * version

=item * type

=item * stage

=back

    use CPANfile::Parse::PPI;
    use Data::Printer;

    my $required = 'requires "CPANfile::Parse::PPI" => 3.6;';
    my $cpanfile = CPANfile::Parse::PPI->new( \$required );
    
    my $modules = $cpanfile->modules;
    p $modules;
    
    __DATA__
    [
        [0] {
            name      "CPANfile::Parse::PPI",
            stage     "",
            type      "requires",
            version   3.6
        }
    ]

=head1 LIMITATIONS

As this is a static parser, this module cannot handle dynamic
code like

    for my $module (qw/
        IO::All
        Zydeco::Lite::App
    /) {
        requires $module, '0';
    }

This module warns when the required "module" doesn't look like
a package name.

You can make it die when you pass C<-strict> to the module
when you load it:

    use CPANfile::Parse::PPI -strict;
    use Data::Printer;

    my $required = do { local $/; <DATA> };
    my $cpanfile = CPANfile::Parse::PPI->new( \$required );
    
    my $modules = $cpanfile->modules;
    
    __DATA__
    for my $module (qw/
        IO::All
        Zydeco::Lite::App
    /) {
        requires $module, '0';
    }
